Deprecate `RequestParametersTrait`

mamazu opened the pull request:
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | yes
| Fixed tickets | -
| Related issues/PRs | #7811 
| License | MIT
| Documentation PR | -

#### What's in this PR?
Deprecating the `RequestParametersTrait`

#### Why?
We want to use the `$request` object correctly and not just get parameters from query and body always. (Technically also a minor performance boost.

#### Example Usage
```php
$request->query->get('param');
$request->headers->get('param');
$request->request->get('param');

// If you need a typesafe alternative (like for booleans)
$request->query->getBoolean('param');
$request->query->getString('param');
$request->query->getInt('param');
```

mamazu committed:
Deprecate `RequestParametersTrait`

mamazu committed:
Deprecate `RequestParametersTrait`

alexander-schranz commented on src/Sulu/Component/Rest/RequestParametersTrait.php:
```diff
@@ -23,6 +23,8 @@ trait RequestParametersTrait
     /**
      * returns request parameter with given name.
      *
+     * @deprecated since version 2.6 will be removed in 3.0 please use the implementation
```
```suggestion
     * @deprecated since version 2.6 will be removed in 3.0 please use $request->query->get() or $request->request->get() or similar type-safe methods instead
```

mamazu commented:
Do we also need a changelog entry?

alexander-schranz commented:
a entry in the `UPGRADE.md` would be nice

mamazu commented:
And maybe also a rector rule to do it for you? :D

alexander-schranz commented on UPGRADE.md:
````diff
@@ -2,6 +2,23 @@
+$request->headers->get('param');
+$request->request->get('param');
+```
+if the parameter does not exist, it throws an exception `Sulu\Component\Rest\Exception\MissingParameterException`.
````
this sentence is strange.

We also should deprecate that exception is somebody catch or relay on it beside ourself in 2.6.

mamazu replied:
Well currently the exception is also mapped to a 400 in the app. I'd still keep it around.

mamazu replied:
`MissingParameterException` is now also marked as deprecated.

alexander-schranz commented on UPGRADE.md:
````diff
@@ -2,6 +2,23 @@
+$request->headers->get('param');
+$request->request->get('param');
+```
+if the parameter does not exist, it throws an exception `Sulu\Component\Rest\Exception\MissingParameterException`.
````
````suggestion
The trait has been deprecated and should be replaced with its implementation. Namely

```php
$request->query->get('param');
$request->headers->get('param');
$request->request->get('param');
```

if the parameter does not exist, it throws an exception `Sulu\Component\Rest\Exception\MissingParameterException`.
````

alexander-schranz commented on UPGRADE.md:
````diff
@@ -2,6 +2,23 @@
+```
+if the parameter does not exist, it throws an exception `Sulu\Component\Rest\Exception\MissingParameterException`.
+
+If you want a typesafe alternative use these functions on all `ParameterBag`s:
+```php
+$request->query->getBoolean('param');
+$request->query->getString('param');
+$request->query->getInt('param');
+```
````
````suggestion
If you want a typesafe alternative use these functions on all `ParameterBag`s:

```php
$request->query->getBoolean('param');
$request->query->getString('param');
$request->query->getInt('param');
```
````

alexander-schranz commented on UPGRADE.md:
```diff
@@ -2,6 +2,23 @@
 
 ## 2.6.7
 
+### Deprecated `RequestParameters` trait
```
```suggestion
### Deprecated `RequestParametersTrait`
```

mamazu commented:
Updated the docs now it should be clearer.

mamazu commented:
@alexander-schranz If you want to have a look again. Then I can go ahead and remove it in 3.0

mamazu commented on UPGRADE-2.x.md:
```diff
@@ -23,6 +23,24 @@ All `.yml` routing files of the core bundles will be removed in the future.
 
 If you want to upgrade your own custom routes, have a look at the opt out command of [RestRoutingBundle](https://github.com/handcraftedinthealps/RestRoutingBundle).
 
+### Deprecated `RequestParameters` trait
```
Should we also mention in this part of the upgrade guide that there is a rector for large code bases to run?

alexander-schranz committed:
Document deprecation of `RequestParameters` trait

Add deprecation notice for the `RequestParameters` trait and provided alternatives for accessing request parameters directly.

alexander-schranz committed:
Merge branch '2.6' into deprecating_request_trait

mamazu commented:
Here is the rector rules for it: https://github.com/mamazu/sulu-rector/blob/65bd0349936834c6f6e4f1d3fb4b05bc19315237/src/Rector/RequestParameterTraitRector.php

alexander-schranz committed:
Fix typo in deprecated trait name

Corrected the name of the deprecated trait from 'RequestParameters' to 'RequestParametersTrait' for clarity.

alexander-schranz committed:
Fix typo in deprecated trait name

Corrected the name of the deprecated trait from 'RequestParameters' to 'RequestParametersTrait' for clarity.

mamazu committed:
Merge branch '2.6' into deprecating_request_trait

mamazu committed:
Removingwithout replacement useless changes
